PEP 729: Typing governance process (#3455)
python/peps: peps/pep-0729.rst

PEP: 729
Title: Typing governance process
Author: Jelle Zijlstra <jelle.zijlstra@gmail.com>, Shantanu Jain <hauntsaninja at gmail.com>
Discussions-To: https://discuss.python.org/t/proposed-new-typing-governance-process/34244
Status: Draft
Type: Process
Topic: Typing

> Content-Type: text/x-rst

Reviewer: Re: https://github.com/python/peps/pull/3454 
```suggestion
```

specification and conformance test suite and will initially be appointed by the Python Steering Council.

Motivation
==========

The Python type system was created by :pep:`484`, now almost ten years ago. The type
system is now widely used, and typing has become an important tool for writing
good, maintainable Python code. Many changes have been made to the type system to cover
more use cases and improve usability. Several type checkers have been created, each
with their own strengths. The type annotation syntax has driven several major innovations
in the Python ecosystem, such as the popular :mod:`dataclasses` module, runtime type
checking and validation by packages such as Pydantic, and static compilation by tools
such as mypyc.

However, as the type system has grown, several interrelated problems with the current
way to manage the type system have become apparent.

PEPs are the only specification
--------------------------------

The Python type system was initially created by a PEP (:pep:`484`), and
changes to the type system are still made by PEPs. The specification for
the Python type system, to the extent there is one, consists of this series

> of PEPs. But PEPs aren't meant to be living documents

Reviewer: [Indeed](https://peps.python.org/pep-0001/#pep-maintenance):

> In general, PEPs are no longer substantially modified after they have reached the Accepted, Final, Rejected or Superseded state. Once resolution is reached, a PEP is considered a historical document rather than a living specification.

But:

> Active (Informational and Process) PEPs may be updated over time to reflect changes to development practices and other details. 

So maybe? 

```suggestion
of PEPs. But Standard Track PEPs aren't meant to be living documents
```

Reviewer: Sure, perhaps this? Or it's also fine as is.

```diff
-But PEPs aren't meant to be living documents
+But PEPs generally aren't meant to be living documents
``

Author: Maybe "these PEPs", referring back to the "series of PEPs" from the previous sentence, all of which are standards-track.

```suggestion
of PEPs. But these PEPs aren't meant to be living documents
```

Author: OK let's do it, but the correct spelling is Standard*s* Track:
```suggestion
of PEPs. But Standards Track PEPs aren't meant to be living documents
```

been instrumental to the success of Python 3. Both modules
have evolved greatly since their initial creation and inspired changes to the
core language.

However, :mod:`asyncio` and :mod:`typing` are different in an essential aspect:
a user who uses :mod:`asyncio` interacts only with the standard library itself,
while a user of :mod:`typing` has to also think about an external tool, the type
checker. The Python language reference covers the symbols in the typing module, but does
not (and should not) go into detail on how the full type system should be
interpreted by type checkers. That material currently exists only in the PEPs.

This problem is shared by the packaging ecosystem, which attempts to solve it
by maintaining a separate set of `PyPA specifications <https://packaging.python.org/en/latest/specifications/>`_.

It's hard to evolve the specification
-------------------------------------

Because the PEPs are the only specification we have, anything that could be seen
as a change to the specification would theoretically require a new PEP. But that
is often too heavy a process for a small change. Sometimes changes are made

> directly to old PEP instead, but that goes against the idea that accepted and

Reviewer: ```suggestion
directly to old PEPs instead, but that goes against the idea that accepted and
```

  argument annotations. Nevertheless, type checkers have long accepted such
  usage.
* A `2023 discussion <https://discuss.python.org/t/pep-561-clarification-regarding-n/32875>`_
  noted that :pep:`561`'s description of partial stubs is unclear, and
  major type checkers did not implement it exactly as specified.
* The widely used third-party ``typing_extensions`` package provides backports of new
  type system features. Type checkers are expected to treat symbols in this
  module the same as symbols in :mod:`typing`, but this is not explicitly
  specified in any of the PEPs.

The type system is underspecified
---------------------------------

While the PEPs provide a specification, they are often not sufficiently precise
(sometimes intentionally so). This is especially true as the combinatorial
complexity of the type system has grown.

It ends up falling to individual type checkers to decide how to navigate
underspecified areas. In cases where type checkers informally coordinate, this
results in de facto standards that aren't clearly recorded anywhere, making
the type system less accessible to newcomers. For example:

* How ``@overload`` matching works

> * How ParamSpec `should work <https://github.com/python/typing/discussions/946>`_ `with methods <https://github.com/microsoft/pyright/issues/3954#issuecomment-1250098464>`_

Reviewer: ```suggestion
* How ``ParamSpec`` `should work <https://github.com/python/typing/discussions/946>`_ `with methods <https://github.com/microsoft/pyright/issues/3954#issuecomment-1250098464>`_
```

  such as runtime type checking, static compilation, IDE support, and documentation.
  The Typing Council should consider all of these use cases when making decisions,
  and be open to supporting additional use cases as they come up.
* **Usable**: The type system should be easy to use for Python developers. It
  should be ergonomic to write well-typed Python code that is accepted by type
  checkers. There should be good documentation for the type system.
* **Stable**: As the type system matures, users should be able to rely on their
  typed code continuing to work and be able to trust their mental model for the
  type system. Changes should be made with care and in a way
  that minimizes disruption. Nevertheless, the type system should be able to
  evolve, and type checkers need not be held to the same strict compatibility
  standards as the Python language itself.

Operations and process
----------------------

The council would have three members, comprised of prominent community members,
such as Python core developers and maintainers of major type checkers. The members
should include people affiliated with a variety of projects related to type checking,

> which may include type checkers, CPython, typeshed, or other projects.

Reviewer: The SC has a limit on the number of people employed by the same company. Does the TC need something similar?

Reviewer 2: In general, thanks to existence of SC, I think we can mostly assume good faith and not have to spell out a bunch of hard requirements for TC. Norm of some diversity in project affiliation should be enough to accomplish TC's goals

Author: Agree, I don't think we need an explicit requirement.

Also, thanks for the review Hugo!

The Steering Council would appoint the initial Typing Council. There is no term
limit for council members. Council members may resign their position at any time.
There is a norm that at least one person will voluntarily resign from the
Typing Council each year.

To determine replacements, nominations will be collected from the typing

> community. Self-nominations are allowed. The Typing Council will then decide

Reviewer 2: ```suggestion
community. Self-nominations are allowed. The existing Typing Council will then decide
```
(to address conversation below)

the replacement member(s) from the nominees. The expectation is that this would
be done by fiat, but the Typing Council can choose a replacement by any means

> they see fit, including a vote.

Reviewer: Do the resigning members participate in selecting the replacements?

Reviewer 2: My intent was that they could, three is a number that helps break ties

to delegate decisions on some PEPs to the Typing Council (exactly as any other
PEP delegation).

Some examples of how past and recent issues could have been handled under this model:

- A PEP like :pep:`695` (type parameter syntax), which changes the language
  syntax, would need to be decided upon by the Steering Council; the Typing
  Council would merely provide opinion or endorsement. Similarly, PEPs
  like :pep:`702` would be decided upon by the Steering
  Council, because it concerns runtime behaviour beyond pure typing. Other examples
  that would need to be decided by the SC include :pep:`718` and :pep:`727`.
- A PEP like :pep:`698` (``@override``), which affects only users of type
  checkers and does not change the overall language, would also by default
  be decided upon by the Steering Council. However, such PEPs could be
  delegated to the Typing Council for a decision (like any other PEP delegation).
  Other examples of PEPs that could potentially be delegated include
  :pep:`647`, :pep:`655`, :pep:`673`, and :pep:`675`.
- Adding a smaller feature, such as :data:`typing.Never` as an alias for
  :data:`typing.NoReturn`, would be done by means of a PR to the spec and

>   conformance  test suite. The Typing

Reviewer: ```suggestion
  conformance test suite. The Typing
```

  Council would then decide whether or not to merge the PR. They may ask for the
  feature to be specified and discussed in a PEP if they feel that is warranted.
- If there is confusion about the interpretation of some part of the spec, like
  happened recently with `partial stubs in PEP
  561 <https://discuss.python.org/t/pep-561-clarification-regarding-n/32875/27>`_,

>   somebody would make a PR to the typing specifications to clarify the

Reviewer: ```suggestion
  somebody would make a PR to the typing specification to clarify the
```

PEPs are the only specification
--------------------------------

The Python type system was initially created by a PEP (:pep:`484`), and
changes to the type system are still made by PEPs. The specification for
the Python type system, to the extent there is one, consists of this series
of PEPs. But PEPs aren't meant to be living documents
or specifications; they are change proposals.

An example may illustrate the problem here. Around the same time
as the introduction of the :mod:`typing` module by :pep:`484`, :pep:`3156`
introduced the :mod:`asyncio` module, another major new feature that has
been instrumental to the success of Python 3. Both modules
have evolved greatly since their initial creation and inspired changes to the
core language.

However, :mod:`asyncio` and :mod:`typing` are different in an essential aspect:
a user who uses :mod:`asyncio` interacts only with the standard library itself,
while a user of :mod:`typing` has to also think about an external tool, the type
checker. The Python language reference covers the symbols in the typing module, but does
not (and should not) go into detail on how the full type system should be

> interpreted by type checkers. That material currently exists only in the PEPs.

Author: The mypy documentation probably comes closest, but it's a user guide, not a reference document to the type system.

that the Typing Council could shepherd.

Mandate
-------

The Typing Council's mandate is to ensure that the Python type system is:

* **Useful**: The type system should serve common use cases. As identified
  by :pep:`484`, the primary use case is static analysis, but there are others,
  such as runtime type checking, static compilation, IDE support, and documentation.
  The Typing Council should consider all of these use cases when making decisions,
  and be open to supporting additional use cases as they come up.
* **Usable**: The type system should be easy to use for Python developers. It
  should be ergonomic to write well-typed Python code that is accepted by type
  checkers. There should be good documentation for the type system.
* **Stable**: As the type system matures, users should be able to rely on their
  typed code continuing to work and be able to trust their mental model for the
  type system. Changes should be made with care and in a way
  that minimizes disruption. Nevertheless, the type system should be able to
  evolve, and type checkers need not be held to the same strict compatibility

>   standards as the Python language itself.

Author: It might be worth expanding on this. Here's a few thoughts:

- The symbols in the typing module should be held to the same standards of compatibility as anything else in the standard library, i.e. we can't just make `typing.Something` disappear at any point, only after we've gone through DperecationWarnings for a few releases.
- But for most of the Council's decisions, the compatibility we're concerned about would be in the behavior of type checkers, and there I don't think it makes sense to use Python's backward compatibility policy. Type checkers make backward-incompatible changes all the time when they tweak their type inference, and they don't follow Python's versioning model.

>   standards as the Python language itself.

Reviewer 3: That sounds about right, with one exception: If there are gray areas where all type checkers currently agree, even if we come up with a proposal for better semantics, I think we ought to be pretty conservative. I believe mypy and pyright (at least) often solve this by keeping the default behavior backwards compatible but offer a flag (sadly type-checker-specific) to enable the "better" approach. For example, mypy started out without `Optional` checking (hard to believe now!) and took a very long time to change the default.

The reason to be conservative in such cases is that if the type checker changes, that often forces maintainers to update their code (if they have policies in place to require passing type checks, which many projects do). I guess you and Shantanu have plenty of experience with how to handle that -- how conservative you should be often depends on how much churn we expect in the community as a result of the change.

>   standards as the Python language itself.

Author: I will push a rewrite of this last sentence, let's see if that works better.

Operations and process
----------------------

> The council would have three members, comprised of prominent community members,

Reviewer 3: Yeah, but I don't think that the EB is slow because we're trying to get too many people involved. We're just all five of us a bit busy. :-)

Maybe you could think who you'd ask to be on the council and discuss your choices with the SC (and with your selected candidates)?

> The council would have three members, comprised of prominent community members,

Author: Good point, I will do that

such as Python core developers and maintainers of major type checkers. The members
should include people affiliated with a variety of projects related to type checking,
which may include type checkers, CPython, typeshed, or other projects.

The Steering Council would appoint the initial Typing Council. There is no term
limit for council members. Council members may resign their position at any time.
There is a norm that at least one person will voluntarily resign from the
Typing Council each year.

To determine replacements, nominations will be collected from the typing
community. Self-nominations are allowed. The existing Typing Council will then decide
the replacement member(s) from the nominees. The expectation is that this would
be done by fiat, but the Typing Council can choose a replacement by any means
they see fit, including a vote.

The Typing Council remains accountable to the Steering Council. At any point,
for any reason, the Steering Council could (publicly or privately) make a
specific change or request a non-specific change to the composition of the

> Typing Council.

Reviewer 2: The SC ultimately has authority over the type system. But I don't think the SC should make changes directly... If SC ever wanted to override TC, I feel that should in the form of a PEP that the SC accepts. Similarly, if the event of irreconcilable TC deadlock about type system, it sounds right that there should be a PEP that the SC decides on.

(Note in this version of the draft that all PEPs go to the SC by default)

> Typing Council.

Reviewer 3: Okay, agreed they should still go through the PEP process. It's interesting though, since the process proposed in this PEP explicitly allows some "minor" changes to the spec to be made without a PEP. I'm guessing that once there's a serious disagreement, by definition it's no longer "minor".

We acknowledge that this is a not particularly democratic structure and puts
a lot of faith in the Typing Council. However, the Python community has a long
history of success with not particularly democratic structures! We believe
self-governance, cycling of membership, and accountability to the
Steering Council will be sufficient to ensure that the Typing Council is meeting

> the needs of the community.

Reviewer 3: Amen.

The council would operate primarily through reviews of GitHub PRs. Regular
meetings are likely not necessary, but the council may set up video calls, a
private chat, or whatever other mechanism they decide upon internally.

The council should aim for transparency, posting all decisions publicly, with a

> rationale if possible.

Reviewer 3: Where? I’d prefer Discourse over GitHub.

Author: Let's say Discourse. Also adding a sentence here saying the Council should wait at least a week to make any decision.

> Members of the council are eligible to sponsor PEPs.

Reviewer 3: This  is a deviation from the current process around sponsorship. Should it be noted in the relevant PEP?

Author: Yes, this is intended as a change proposal, so if this PEP is accepted, the relevant PEP (PEP 1 I think) should be amended. Let's make that explicit:

```suggestion
Members of the council will be eligible to sponsor PEPs. If this PEP is accepted,
:pep:`1` should be amended to note this fact.
```

:func:`typing.get_overloads` introspection helper. As this function does not
affect type checker behavior, it would not require approval by the Typing
Council.

Relationship with type checkers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The Typing Council has no direct authority over type checkers; it cannot
force them to implement particular features or make behavior changes. Type
checkers are incentivized to follow the specification set out by the Council
because it allows them to take advantage of shared resources, such as
libraries that expose typing information that follows the specification,
the stub files in typeshed, the ``typing`` standard library module, and
user documentation that covers the standard type system.
Type checkers are free to extend the type system or deviate from the
specification, but they should document such differences clearly.

The fact that type checkers need to implement any decisions made by the
Typing Council acts as a useful brake on the Council, ensuring that its
decisions are conservative and well-considered. Individual type checkers
remain free to innovate as they see fit, and successful innovations can

> be incorporated into the standard type system.

Reviewer 3: I really like this thoughtful section.

Projects
--------

Here are some efforts a Typing Council would be responsible for.

> Conformance test suite

Reviewer 3: This feels like a Sisyphean task. Aren’t we over-promising a bit?

Author: I think there's value in this even if it what ends up happening is that we rarely add to the test suite except when a PEP comes along and wants to add something to the type system. At the Typing Summit last year there was quite a bit of support for the idea of a conformance test suite.

Reviewer 3: Okay, no worries. As long as we can convince people to update the test suite for new PEPs. (Which may be a problem until there's at least one complete implementation -- you can't really write tests for a feature that's not yet implemented.)

^^^^^^^^^^^^^^^^^^^^^^

A conformance test suite would provide machine checkable documentation for how
type checkers should check Python code, accompanied by the results of major type
checker implementations on the test suite. A rough sketch for what this could
look like was `created by Shantanu <https://github.com/hauntsaninja/type_checker_consistency>`_.

This would contain prescriptive tests from behavior prescribed by previous PEPs
and descriptive tests that let us document behavior of existing implementations
in areas that are not prescribed by any standard. These descriptions would be
useful to inform efforts below and to identify areas of focus for
standardization.

> Specification for the type system

Reviewer 3: I also worry about the size of this task. E.g. seeing Kevin’s spec of just a part of the type system (arguably the easy part :-) makes me worry. Not that I have a better idea…

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A specification would initially be created by stitching together the
specification sections from the existing PEPs, and then gradually improved to
clarify points of confusion and cover more areas. A draft of such a

> stitched-together spec was `created by Jelle <https://github.com/JelleZijlstra/typing-spec>`_.

Reviewer 3: Aren’t PEP 483/484 way too imprecise for this purpose?

Author: Arguably yes, but it's what we currently have. This proposal builds a foundation that then allows the community to work towards a more formalized spec. As you wrote a little way up, building a whole spec from scratch is a Sisyphean task, so I think this is the most realistic way to make progress. Once the stitched-together spec is in place, the Council can make amendments to improve it.

Author: Arguably yes, but it's what we currently have. This proposal builds a foundation that then allows the community to work towards a more formalized spec. As you wrote a little way up, building a whole spec from scratch is a Sisyphean task, so I think this is the most realistic way to make progress. Once the stitched-together spec is in place, the Council can make amendments to improve it.

Reviewer 4: I know I've said this elsewhere, but we don't need to build from scratch to be successful here, nor is the stitched-up way necessarily best. We could quickly build this up from theory (I know some people would prefer not to) and check for differences to current type checking behavior, note these, then decide what to do about reconciling them, choosing current behavior, modifying definitions to retain a specific behavior, and so on.

This gets us a "best of both worlds" in a good specification starting point quickly from theory, but grounded against the current landscape to examine how to proceed.

Author: That doesn't sound very quick! I'm not sure how what you're suggesting is different from starting from scratch. It will likely require quite a bit of work to get alignment from the different type checkers.

The PEP has a "Rejected ideas" section for writing the spec from scratch.

Reviewer 4: > I'm not sure how what you're suggesting is different from starting from scratch.

Starting with any choice of well-defined theory, the theory already exists, we could even rather readily compare with multiple models here, none of which would need to be from scratch. 

We then look for where type checkers deviate from theory. This can be automated against the test suite already because the theories are machine provable, so we can feed them something and compare what they say with what type checkers currently say.

<Details><Summary> program in coq that was provided in another related discussion which illustrates this capability for one model of typing.</Summary>

started here: https://gist.github.com/mniip/0bfb52f35da3ad6c96878082e6af37e6

```coq
Require Import Coq.Lists.List.
Require Import Coq.Bool.Bool.
Require Import Coq.Relations.Relations.
Require Import Coq.Compat.AdmitAxiom.

Import ListNotations.
Infix "<$>" := map (at level 65, left associativity).
Notation "k >>= f" := (flat_map f k) (at level 66, left associativity).

Definition allb {A} (f : A -> bool) (xs : list A) : bool :=
  fold_right andb true (map f xs).

Definition anyb {A} (f : A -> bool) (xs : list A) : bool :=
  fold_right orb false (map f xs).

Definition apList {A B} (fs : list (A -> B)) (xs : list A) : list B :=
  fs >>= fun f => f <$> xs.

Infix "<*>" := apList (at level 65, left associativity).

Definition null {A} (xs : list A) : bool :=
  match xs with
  | [] => true
  | _ => false
  end.

(* Subtyping decision procedure for base types. We parameterize over it here. *)
Class SubtypeDecidable (BaseType : Type) : Type :=
  { bsubDecide : list BaseType -> list BaseType -> bool
    (* `bsubDecide xs ys` is whether the intersection of xs lies within the union of ys *)
  ; BSub := fun t s => bsubDecide [t] [s] = true
  ; bsub_refl : reflexive _ BSub
  ; bsub_trans : transitive _ BSub
  }.

Section Types.

Context {BaseType : Type} `{SubtypeDecidable BaseType}.

Inductive SType : Type :=
  | SBottom : SType
  | STop : SType
  | SNot : SType -> SType
  | SArrow : SType -> SType -> SType
  | SBase : BaseType -> SType
  | SAnd : SType -> SType -> SType
  | SOr : SType -> SType -> SType.

Inductive GType : Type :=
  | GBottom : GType
  | GTop : GType
  | GAny : GType
  | GNot : SType -> GType
  | GArrow : GType -> GType -> GType
  | GBase : BaseType -> GType
  | GAnd : GType -> GType -> GType
  | GOr : GType -> GType -> GType.

Declare Scope stype_scope.
Delimit Scope stype_scope with stype.
Bind Scope stype_scope with SType.
Notation "𝟘" := SBottom : stype_scope.
Notation "𝟙" := STop : stype_scope.
Notation "¬ t" := (SNot t) (at level 75) : stype_scope.
Infix "→" := SArrow (at level 99, right associativity) : stype_scope.
Infix "∧" := SAnd (at level 85, right associativity) : stype_scope.
Infix "∨" := SOr (at level 85, right associativity) : stype_scope.
Notation "⋀ xs" := (fold_right SAnd STop xs) (at level 85) : stype_scope.
Notation "⋁ xs" := (fold_right SOr SBottom xs) (at level 85) : stype_scope.

Declare Scope gtype_scope.
Delimit Scope gtype_scope with gtype.
Bind Scope gtype_scope with GType.
Notation "𝟘" := GBottom : gtype_scope.
Notation "𝟙" := GTop : gtype_scope.
Notation "?" := GAny : gtype_scope.
Notation "¬ t" := (GNot t) (at level 75) : gtype_scope.
Infix "→" := GArrow (at level 99, right associativity) : gtype_scope.
Infix "∧" := GAnd (at level 85, right associativity) : gtype_scope.
Infix "∨" := GOr (at level 85, right associativity) : gtype_scope.
Notation "⋀ xs" := (fold_right GAnd GTop xs) (at level 85) : gtype_scope.
Notation "⋁ xs" := (fold_right GOr GBottom xs) (at level 85) : gtype_scope.

Reserved Infix "≤" (at level 70, no associativity).
Inductive Sub : SType -> SType -> Prop :=
  | sub_base : forall b c, BSub b c -> SBase b ≤ SBase c
  | sub_not_mono : forall t s, s ≤ t -> (¬t) ≤ (¬s)
  | sub_arrow_mono : forall t1 t2 s1 s2, s1 ≤ t1 -> t2 ≤ s2 -> (t1 → t2) ≤ (s1 → s2)
  | sub_bottom : forall t, 𝟘 ≤ t
  | sub_top : forall t, t ≤ 𝟙
  | sub_and_l : forall t1 t2, (t1 ∧ t2) ≤ t1
  | sub_and_r : forall t1 t2, (t1 ∧ t2) ≤ t2
  | sub_and : forall t s1 s2, t ≤ s1 -> t ≤ s2 -> t ≤ (s1 ∧ s2)
  | sub_or_l : forall t1 t2, t1 ≤ (t1 ∨ t2)
  | sub_or_r : forall t1 t2, t2 ≤ (t1 ∨ t2)
  | sub_or : forall t1 t2 s, t1 ≤ s -> t2 ≤ s -> (t1 ∨ t2) ≤ s
  | sub_trans : forall t s r, t ≤ s -> s ≤ r -> t ≤ r
where "t1 ≤ t2" := (Sub t1 t2).

Section SubDecide.

(* Either conjuncts or disjuncts depending on context, categorized by whether
it's a BaseType or an arrow, and by whether it's negated *)
Record Junct := mkJunct
  { bases : list BaseType
  ; arrows : list (SType * SType)
  ; notBases : list BaseType
  ; notArrows : list (SType * SType)
  }.

Definition unitJunct : Junct :=
  {| bases := []
  ; arrows := []
  ; notBases := []
  ; notArrows := []
  |}.

Definition mergeJuncts (j1 j2 : Junct) : Junct :=
  {| bases := bases j1 ++ bases j2
  ; arrows := arrows j1 ++ arrows j2
  ; notBases := notBases j1 ++ notBases j2
  ; notArrows := notArrows j1 ++ notArrows j2
  |}.

Definition invertJunct (j : Junct) : Junct :=
  {| bases := notBases j
  ; arrows := notArrows j
  ; notBases := bases j
  ; notArrows := notArrows j
  |}.

Fixpoint udnf (t : SType) : list Junct :=
  match t with
  | SBase b => [{| bases := [b]; arrows := []; notBases := []; notArrows := [] |}]
  | t1 → t2 => [{| bases := []; arrows := [(t1, t2)]; notBases := []; notArrows := [] |}]
  | 𝟘 => []
  | 𝟙 => [unitJunct]
  | ¬t => invertJunct <$> ucnf t
  | t1 ∨ t2 => udnf t1 ++ udnf t2
  | t1 ∧ t2 => mergeJuncts <$> udnf t1 <*> udnf t2
  end%stype
with ucnf (t : SType) : list Junct :=
  match t with
  | SBase b => [{| bases := [b]; arrows := []; notBases := []; notArrows := [] |}]
  | t1 → t2 => [{| bases := []; arrows := [(t1, t2)]; notBases := []; notArrows := [] |}]
  | 𝟘 => [unitJunct]
  | 𝟙 => []
  | ¬t => invertJunct <$> udnf t
  | t1 ∧ t2 => ucnf t1 ++ ucnf t2
  | t1 ∨ t2 => mergeJuncts <$> ucnf t1 <*> ucnf t2
  end%stype.

Definition junctDecide (jt js : Junct) : bool :=
  match bases jt ++ notBases js, arrows jt ++ notArrows js with
  | lbases, [] => bsubDecide lbases (bases js ++ notBases jt)
  | [], _ => match proof_admitted with
             (* How to decide subtyping of a bunch of arrows? Not sure. *)
             end
  | _, _ => match proof_admitted with
            (* How to decide whether base types intersect arrow types?
            This depends on the target language in some way. *)
            end
  end.

Definition subDecide (t s : SType) : bool :=
  allb id (junctDecide <$> udnf t <*> ucnf s).

End SubDecide.

Infix "≤!" := subDecide (at level 70, no associativity).

Notation "t ∈ S" := (S t) (at level 70, only parsing).

Fixpoint γ (τ : GType) : SType -> Prop :=
  match τ with
  | ? => fun _ => True
  | 𝟘 => eq 𝟘%stype
  | 𝟙 => eq 𝟙%stype
  | GBase b => eq (SBase b)
  | ¬t => eq (¬t)%stype
  | τ1 ∧ τ2 => fun t => match t with
                        | t1 ∧ t2 => t1 ∈ γ τ1 /\ t2 ∈ γ τ2
                        | _ => False
                        end%stype
  | τ1 ∨ τ2 => fun t => match t with
                        | t1 ∨ t2 => t1 ∈ γ τ1 /\ t2 ∈ γ τ2
                        | _ => False
                        end%stype
  | τ1 → τ2 => fun t => match t with
                        | t1 → t2 => t1 ∈ γ τ1 /\ t2 ∈ γ τ2
                        | _ => False
                        end%stype
  end%gtype.

Definition lift2 (P : SType -> SType -> Prop) : GType -> GType -> Prop
  := fun τ1 τ2 => exists t1 t2, t1 ∈ γ τ1 /\ t2 ∈ γ τ2 /\ P t1 t2.

Fixpoint gradualMinimum (τ : GType) : SType :=
  match τ with
  | ? => 𝟘%stype
  | 𝟘 => 𝟘%stype
  | 𝟙 => 𝟙%stype
  | GBase b => SBase b
  | ¬t => (¬t)%stype
  | τ1 ∧ τ2 => (gradualMinimum τ1 ∧ gradualMinimum τ2)%stype
  | τ1 ∨ τ2 => (gradualMinimum τ1 ∨ gradualMinimum τ2)%stype
  | τ1 → τ2 => (gradualMaximum τ1 → gradualMinimum τ2)%stype
  end%gtype
with gradualMaximum (τ : GType) : SType :=
  match τ with
  | ? => 𝟙%stype
  | 𝟘 => 𝟘%stype
  | 𝟙 => 𝟙%stype
  | GBase b => SBase b
  | ¬t => (¬t)%stype
  | τ1 ∧ τ2 => (gradualMaximum τ1 ∧ gradualMaximum τ2)%stype
  | τ1 ∨ τ2 => (gradualMaximum τ1 ∨ gradualMaximum τ2)%stype
  | τ1 → τ2 => (gradualMinimum τ1 → gradualMaximum τ2)%stype
  end%gtype.

Notation "τ ⇓" := (gradualMinimum τ) (at level 65).
Notation "τ ⇑" := (gradualMaximum τ) (at level 65).

Infix "̃≤" := (lift2 Sub) (at level 70, no associativity).

Definition consSubDecide (σ τ : GType) : bool := σ⇓ ≤! τ⇑.

Infix "̃≤!" := consSubDecide (at level 70, no associativity).

Infix "̃≰" := (lift2 (fun s t => ~(s ≤ t))) (at level 70, no associativity).

Definition consNotSubDecide (σ τ : GType) : bool := negb (σ⇑ ≤! τ⇓).

Infix "̃≰!" := consNotSubDecide (at level 70, no associativity).

(* ??? The paper completely glances over that it treats SType as a subset of
GType in the definition of γA. γA+ restricted to SType is γApos' here. *)
Fixpoint stype2gtype (t : SType) : GType :=
  match t with
  | 𝟘 => 𝟘%gtype
  | 𝟙 => 𝟙%gtype
  | ¬t => (¬t)%gtype
  | t1 → t2 => (stype2gtype t1 → stype2gtype t2)%gtype
  | SBase b => GBase b
  | t1 ∧ t2 => (stype2gtype t1 ∧ stype2gtype t2)%gtype
  | t1 ∨ t2 => (stype2gtype t1 ∨ stype2gtype t2)%gtype
  end%stype.

Fixpoint γAneg (t : SType) : list (list (GType * GType)) :=
  match t with
  | t1 ∨ t2 => (fun T1 T2 => T1 ++ T2) <$> γAneg t1 <*> γAneg t2
  | t1 ∧ t2 => γAneg t1 ++ γAneg t2
  | _ → _ => [[]]
  | ¬t => γApos' t
  | 𝟘 => [[]]
  | 𝟙 => []
  | SBase _ => [[]]
  end%stype
with γApos' (t : SType) : list (list (GType * GType)) :=
  match t with
  | t1 ∨ t2 => γApos' t1 ++ γApos' t2
  | t1 ∧ t2 => (fun T1 T2 => T1 ++ T2) <$> γApos' t1 <*> γApos' t2
  | t1 → t2 => [[(stype2gtype t1, stype2gtype t2)]]
  | ¬t => γAneg t
  | 𝟘 => []
  | 𝟙 => [[]]
  | SBase _ => [[]]
  end%stype.

Fixpoint γApos (τ : GType) : list (list (GType * GType)) :=
  match τ with
  | τ1 ∨ τ2 => γApos τ1 ++ γApos τ2
  | τ1 ∧ τ2 => (fun T1 T2 => T1 ++ T2) <$> γApos τ1 <*> γApos τ2
  | τ1 → τ2 => [[(τ1, τ2)]]
  | ¬t => γAneg t
  | 𝟘 => []
  | 𝟙 => [[]]
  | GBase _ => [[]]
  | ? => [[(?, ?)]]
  end%gtype.

Definition dom (τ : GType) : SType :=
  ⋀ (fun S => ⋁ (fun '(ρ, _) => ρ⇑) <$> S)%stype <$> γApos τ.

Fixpoint assignments {X : Type} (xs : list X) : list (list (X * bool)) :=
  match xs with
  | [] => [[]]
  | (x :: xs) => cons <$> [(x, false); (x, true)] <*> assignments xs
  end.

Definition partitions {X : Type} (xs : list X) : list (list X * list X) :=
  (fun ys => (fst <$> filter snd ys, fst <$> filter (fun p => negb (snd p)) ys))
  <$> assignments xs.

Definition appResult (τ σ : GType) : GType := ⋁ (fun S =>
  ⋁ (fun '(Q, Qbar) => ⋀ snd <$> Qbar) <$> filter
    (fun '(Q, Qbar) => negb (null Qbar) && (σ ̃≰! ⋁ fst <$> Q)
      && negb ((σ⇑ ∧ ⋁ (fun '(ρ, _) => ρ⇑) <$> Qbar) ≤! 𝟘))
    (partitions S)
  )%gtype <$> γApos τ.

Infix "∘" := appResult (at level 90, no associativity).

Section SomeTheorems.

Local Theorem sub_and_mono t1 t2 s1 s2 : t1 ≤ s1 -> t2 ≤ s2 -> (t1 ∧ t2) ≤ (s1 ∧ s2).
Proof.
  intros p q.
  apply sub_and.
  - refine (sub_trans _ _ _ _ p).
    apply sub_and_l.
  - refine (sub_trans _ _ _ _ q).
    apply sub_and_r.
Qed.

Local Theorem sub_or_mono t1 t2 s1 s2 : t1 ≤ s1 -> t2 ≤ s2 -> (t1 ∨ t2) ≤ (s1 ∨ s2).
Proof.
  intros p q.
  apply sub_or.
  - refine (sub_trans _ _ _ p _).
    apply sub_or_l.
  - refine (sub_trans _ _ _ q _).
    apply sub_or_r.
Qed.

Local Theorem sub_refl t : t ≤ t.
Proof.
  pose proof sub_and_mono.
  pose proof sub_or_mono.
  pose proof bsub_refl.
  induction t; auto; now constructor.
Qed.

Local Theorem gradual_extrema_in_concretization τ : τ⇓ ∈ γ τ /\ τ⇑ ∈ γ τ.
Proof.
  induction τ as [ | | | | ? [] ? [] | | ? [] ? [] | ? [] ? [] ]; simpl; auto.
Qed.

Local Theorem gradual_extrema_bounds τ : forall t, t ∈ γ τ -> τ⇓ ≤ t /\ t ≤ τ⇑.
Proof.
  induction τ as [ | | | | ? IHl ? IHr | | ? IHl ? IHr | ? IHl ? IHr ]; simpl.
  - intros _ <-. repeat constructor.
  - intros _ <-. repeat constructor.
  - repeat constructor.
  - intros _ <-. repeat constructor.
    + apply sub_refl.
    + apply sub_refl.
  - intros [ | | | t1 t2 | | | ]; try contradiction.
    intros [].
    destruct (IHl t1); auto.
    destruct (IHr t2); auto.
    repeat constructor; auto.
  - intros _ <-. repeat constructor.
    + apply bsub_refl.
    + apply bsub_refl.
  - intros [ | | | | | t1 t2 | ]; try contradiction.
    intros [].
    destruct (IHl t1); auto.
    destruct (IHr t2); auto.
    split.
    + apply sub_and_mono; auto.
    + apply sub_and_mono; auto.
  - intros [ | | | | | | t1 t2 ]; try contradiction.
    intros [].
    destruct (IHl t1); auto.
    destruct (IHr t2); auto.
    split.
    + apply sub_or_mono; auto.
    + apply sub_or_mono; auto.
Qed.

Local Theorem cons_sub_reduction σ τ : σ ̃≤ τ <-> σ⇓ ≤ τ⇑.
Proof.
  unfold lift2. split.
  - intros [s [t [? []]]].
    apply (sub_trans _ s _).
    + now apply gradual_extrema_bounds.
    + apply (sub_trans _ t _).
      * assumption.
      * now apply gradual_extrema_bounds.
  - intros ?.
    exists (σ⇓), (τ⇑).
    split; try split.
    + apply gradual_extrema_in_concretization.
    + apply gradual_extrema_in_concretization.
    + assumption.
Qed.

Theorem sub_decide_agrees t s : t ≤ s <-> t ≤! s = true.
  (* Depends on the contents of junctDecide *)
Admitted.

Theorem cons_sub_decide_agrees σ τ : σ ̃≤ τ <-> σ ̃≤! τ = true.
Proof.
  transitivity (σ⇓ ≤ τ⇑).
  - apply cons_sub_reduction.
  - unfold consSubDecide.
    apply sub_decide_agrees.
Qed.

Theorem cons_not_sub_reduction σ τ : σ ̃≰ τ <-> ~(σ⇑ ≤ τ⇓).
Proof.
  unfold lift2. split.
  - intros [s [t [? [? p]]]] n.
    apply p.
    apply (sub_trans _ (σ⇑) _).
    + now apply gradual_extrema_bounds.
    + apply (sub_trans _ (τ⇓) _).
      * assumption.
      * now apply gradual_extrema_bounds.
  - intros n.
    exists (σ⇑), (τ⇓).
    split; try split.
    + apply gradual_extrema_in_concretization.
    + apply gradual_extrema_in_concretization.
    + assumption.
Qed.

Theorem cons_not_sub_decide_agrees σ τ : σ ̃≰ τ <-> σ ̃≰! τ = true.
Proof.
  transitivity (~(σ⇑ ≤ τ⇓)).
  - apply cons_not_sub_reduction.
  - unfold consNotSubDecide.
    transitivity (σ⇑ ≤! τ⇓ <> true).
    + apply not_iff_compat.
      apply sub_decide_agrees.
    + transitivity (σ⇑ ≤! τ⇓ = false).
      * apply not_true_iff_false.
      * symmetry. apply negb_true_iff.
Qed.

End SomeTheorems.

Section SomeContradictions.

Inductive PartialEvalResult : Type :=
  | EvBottom
  | EvUnknown
  | EvTop.

Local Fixpoint partialEval (t : SType) : PartialEvalResult :=
  match t with
  | 𝟘 => EvBottom
  | 𝟙 => EvTop
  | SBase _ => EvUnknown
  | _ → _ => EvUnknown
  | ¬t => match partialEval t with
          | EvBottom => EvTop
          | EvUnknown => EvUnknown
          | EvTop => EvBottom
          end
  | t1 ∧ t2 => match partialEval t1, partialEval t2 with
               | EvBottom, _ => EvBottom
               | _, EvBottom => EvBottom
               | EvUnknown, _ => EvUnknown
               | _, EvUnknown => EvUnknown
               | _, _ => EvTop
               end
  | t1 ∨ t2 => match partialEval t1, partialEval t2 with
               | EvTop, _ => EvTop
               | _, EvTop => EvTop
               | EvUnknown, _ => EvUnknown
               | _, EvUnknown => EvUnknown
               | _, _ => EvBottom
               end
  end%stype.

Local Theorem partial_eval_mono t s :
  t ≤ s -> match partialEval t, partialEval s with
           | EvBottom, _ => True
           | _, EvTop => True
           | EvUnknown, EvBottom => False
           | EvUnknown, _ => True
           | EvTop, _ => False
           end.
Proof.
  intros p.
  induction p.
  all: repeat first [ progress (simpl; auto) | destruct (partialEval _) ].
Qed.

Local Theorem top_nle_bottom : ~(𝟙 ≤ 𝟘).
Proof.
  intros n. apply (partial_eval_mono _ _ n).
Qed.

Theorem cons_sub_nontrans : ~(forall g h k, g ̃≤ h -> h ̃≤ k -> g ̃≤ k).
Proof.
  intros n.
  enough (p : 𝟙 ̃≤ 𝟘).
  - apply top_nle_bottom.
    unfold lift2 in p. simpl in p.
    now destruct p as [? [? [-> [->]]]].
  - apply (n _ ?%gtype _).
    + unfold lift2. simpl.
      exists 𝟙%stype, 𝟙%stype.
      split; auto. split; auto.
      constructor.
    + unfold lift2. simpl.
      exists 𝟘%stype, 𝟘%stype.
      split; auto. split; auto.
      constructor.
Qed.

End SomeContradictions.

End Types.

(* Have to redeclare all notations now that the section is closed *)
Declare Scope stype_scope.
Delimit Scope stype_scope with stype.
Bind Scope stype_scope with SType.
Notation "𝟘" := SBottom : stype_scope.
Notation "𝟙" := STop : stype_scope.
Notation "¬ t" := (SNot t) (at level 75) : stype_scope.
Infix "→" := SArrow (at level 99, right associativity) : stype_scope.
Infix "∧" := SAnd (at level 85, right associativity) : stype_scope.
Infix "∨" := SOr (at level 85, right associativity) : stype_scope.
Notation "⋀ xs" := (fold_right SAnd STop xs) (at level 85) : stype_scope.
Notation "⋁ xs" := (fold_right SOr SBottom xs) (at level 85) : stype_scope.
Declare Scope gtype_scope.
Delimit Scope gtype_scope with gtype.
Bind Scope gtype_scope with GType.
Notation "𝟘" := GBottom : gtype_scope.
Notation "𝟙" := GTop : gtype_scope.
Notation "?" := GAny : gtype_scope.
Notation "¬ t" := (GNot t) (at level 75) : gtype_scope.
Infix "→" := GArrow (at level 99, right associativity) : gtype_scope.
Infix "∧" := GAnd (at level 85, right associativity) : gtype_scope.
Infix "∨" := GOr (at level 85, right associativity) : gtype_scope.
Notation "⋀ xs" := (fold_right GAnd GTop xs) (at level 85) : gtype_scope.
Notation "⋁ xs" := (fold_right GOr GBottom xs) (at level 85) : gtype_scope.
Infix "≤" := Sub (at level 70, no associativity).
Infix "≤!" := subDecide (at level 70, no associativity).
Notation "τ ⇓" := (gradualMinimum τ) (at level 65).
Notation "τ ⇑" := (gradualMaximum τ) (at level 65).
Infix "̃≤" := (lift2 Sub) (at level 70, no associativity).
Infix "̃≤!" := consSubDecide (at level 70, no associativity).
Infix "̃≰" := (lift2 (fun s t => ~(s ≤ t))) (at level 70, no associativity).
Infix "̃≰!" := consNotSubDecide (at level 70, no associativity).
Infix "∘" := appResult (at level 90, no associativity).

Section Example.

Inductive BaseType :=
  | object
  | int
  | str
  | bytes
  | float.
Scheme Equality for BaseType.

Definition SBase' := @SBase BaseType.
Coercion SBase' : BaseType >-> SType.
Definition GBase' := @GBase BaseType.
Coercion GBase' : BaseType >-> GType.

Definition decideSubclass (b c : BaseType) : bool :=
  BaseType_beq b c || match b, c with
                      | _, object => true
                      | _, _ => false
                      end.

Program Instance subtype_decidable_basetype : SubtypeDecidable BaseType :=
  {| bsubDecide := fun conjs disjs =>
    anyb id (decideSubclass <$> conjs <*> disjs)
    || anyb (fun c => match c with object => true | _ => false end) disjs
    || (anyb (fun c => match c with int => true | _ => false end) conjs
        && anyb (fun c => match c with str => true | _ => false end) conjs)
  |}.
Next Obligation.
  unfold reflexive.
  intros []; simpl; auto.
Qed.
Next Obligation.
  unfold transitive, anyb.
  intros b c d.
  destruct b, d; simpl; auto; destruct c; simpl; auto.
Qed.

Eval cbn in (int → float) ∧ (str → bytes) ∘ int.
(*
     = (((float ∧ bytes ∧ 𝟙) ∨ (float ∧ 𝟙) ∨ 𝟘) ∨ 𝟘)%gtype
     : GType
*)

End Example.
```

</Details>

One reason I think this approach has a little more usefulness in terms of "get[ting] alignment from the different type checkers." is that type checkers are already deviating from PEPs in some cases, whether because they don't implement a check or behavior that is specified, bugs, and so on. 

Starting by comparing the abstract ideal from where we are currently gives us a good place to start in terms of what to consider for what needs reconciling, where there are current differences, and if we want to specify exceptions due to pragmatism, fix something which has larger impact, or any other decision we could make with the facts available to us.

If we instead start by stitching the PEPs together, we have the "current state but worse", because much of type checking beahvior was never specifyied in PEPs (such as subclassing Any being fine to make mocks work 🙄 ), and that which was adopted in PEPs is acknowledged here as underspecified. We get a very real-world check for the "current state" by just seeing where the current state deviates from theory, and this does not need to be slow like some other formalization efforts have been, because this isn't a formalization, just an observation of where what we have diverges from theory. We don't have to answer hard questions about *should* python match those theories here, just use them as a tool for quickly observing the current state.

Author: Thanks. I think an effort like your is valuable, but it doesn't seem like it can replace the proposed stitched-together spec: it doesn't explain how TypedDicts or ParamSpecs work, or many other features beyond the very basic parts of the type system. Once there is a Typing Council and the initial spec is in place (which I acknowledge will be imperfect), the Council can pursue efforts to formalize parts of the type system. This PEP should primarily focus on getting the Council in place so we have a better decision maker on these foundational questions about the type system.

It's fine if you disagree with that approach, but I'm still convinced it's the right one, and therefore that's what the PEP will propose at the moment. Further discussion about this can happen once there is a Discuss thread about the PEP.

Reviewer 4: I think we differ a bit on what we think this method could accomplish, however, I'm fine with revisiting this later if even with this offered as a potential path you believe that the stitched-together peps have more immediate usefulness. I'll table this until we're looking to move on from the stitched peps to a formal specification, as I believe this approach will offer a similarly useful lens then too.

The specification has a few audiences:

* For type checkers, it provides a description of how an idealized type checker
  should behave. Individual type checkers have different goals and technical
  constraints and they are free to deviate from the spec if they do not have the
  resources to fully implement it or if they believe a different behavior better
  serves their users. However, they should document such deviations from the
  spec.
* For projects such as typeshed, or libraries that want to be compatible with
  multiple type checkers, it provides a set of rules that they can follow to
  make their code understood by type checkers.
* For people who want to propose changes to the type system, it provides a
  foundation for any new proposals.

> Notably, the specification is not aimed at end users of typing. End users

Reviewer 3: Hm. But library maintainers mentioned above are also just end users.

Author: Rewording this to say it's about users who don't need to worry about multiple type checkers (which is likely most normal developers).

  of existing type checkers. If existing type checkers
  behave roughly similarly, that is evidence that their shared
  behavior should be made part of the specification.

User-facing reference for the type system
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Documentation is important for the success of the Python type system, so
the Typing Council should ensure that there is good documentation for the
type system.

As mentioned previously, PEPs are point in time change proposals aimed at
multiple audiences that are hard to clarify. This makes them ill-suited as user
documentation. The specification discussed in the previous section would
be a living document, but it would likely be too technical to serve as
documentation for normal usage.

Therefore, a separate user-facing reference for the type system would be
useful. Such an effort could expand the documentation on
`typing.readthedocs.io <https://typing.readthedocs.io/en/latest/>`_ and
reuse material from the documentation sections of individual type checkers
and the CPython documentation.

Amendments
----------

This PEP serves as a charter for the Typing Council. Changes to its operation

> can be made either through a new PEP or through a change to this PEP. In either

Reviewer 3: I would prefer updating this PEP. When would we need a new PEP?

Author: I suppose if someone in the future decides that this whole Typing Council is a terrible idea and we should throw it away and do something totally different. I hope that's unlikely, but it's worth acknowledging as a possibility.

Reviewer 3: Sure, but that applies to *anything* codified in a PEP. I'm fine with keeping the wording you have now though.

whether the specification accurately reflects current behavior, and if not,
whether the specification or the type checkers should change.

Starting with the existing PEPs creates a lower-quality specification,
but it means that the Typing Council can immediately start making a difference
anywhere in the type system by improving and clarifying the specification.
A formalization effort can still proceed by gradually replacing sections of the
specification.

Alternate governance mechanisms
-------------------------------

An earlier draft of this PEP suggested that the Steering Council appoint
members of the Typing Council each year. The current Steering Council suggested
that it would be better to have the Typing Council self-organise and avoid
the need for the Steering Council to continuously supervise the Typing Council.

Alternate governance mechanisms are possible, including more democratic ones,
but these typically raise several thorny questions, require much heavier
amounts of process and are potentially more divisive. For example, see the PEP
8000 series, or recent discussions about alternative governance in other Python

> subcommunities. Ultimately, the Typing Council exists at the behest of the

Reviewer 3: Is “behest” really the right word here?

Reviewer 2: ```suggestion
subcommunities. Ultimately, the Typing Council exists under the authority of the
```

* For people who want to propose changes to the type system, it provides a
  foundation for any new proposals.

> Notably, the specification is not aimed at end users of typing who do not

Reviewer 3: I'd drop "end", otherwise it's dangerously close to the different meaning you'd get by adding a comma before "who".

As someone (not a typical user :-) who tends to care both about getting VS Code to show "no problems" *and* getting `mypy --strict` to pass I'm still skeptical about whether typical users care about multiple type checkers. Library maintainers probably have to, but they're hopefully not considered to be end users.

Author: Reworded this a bit more.
